mat/dense_arithmetic.go

```diff
@@ -457,67 +457,162 @@ func strictCopy(m *Dense, a Matrix) {
 
 // Exp calculates the exponential of the matrix a, e^a, placing the result
 // in the receiver. Exp will panic with matrix.ErrShape if a is not square.
-//
-// Exp uses the scaling and squaring method described in section 3 of
-// http://www.cs.cornell.edu/cv/researchpdf/19ways+.pdf.
 func (m *Dense) Exp(a Matrix) {
+	// The implementation used here is from Functions of Matrices: Theory and Computation
+	// Chapter 10, Algorithm 10.20. https://doi.org/10.1137/1.9780898717778.ch10
+
 	r, c := a.Dims()
 	if r != c {
 		panic(ErrShape)
 	}
 
-	var w *Dense
-	if m.IsZero() {
-		m.reuseAsZeroed(r, r)
-		w = m
-	} else {
-		w = getWorkspace(r, r, true)
-	}
-	for i := 0; i < r*r; i += r + 1 {
-		w.mat.Data[i] = 1
+	m.reuseAs(r, r)
+	if r == 1 {
+		m.mat.Data[0] = math.Exp(a.At(0, 0))
+		return
 	}
 
-	const (
-		terms   = 10
-		scaling = 4
-	)
-
-	small := getWorkspace(r, r, false)
-	small.Scale(math.Pow(2, -scaling), a)
-	power := getWorkspace(r, r, false)
-	power.Copy(small)
+	pade := []struct {
```

Comment:
perhaps consider lifting this outside of `Expm` ?

Reply:
I'll run benchmarks to see if the compiler puts these in a read only segment/doesn't reallocate, but I'd rather that stayed within the function scope if possible for clarity.

If the compiler *does* reallocate on each call even though the data are not mutated, I think that's a worthwhile issue against golang/go.

Reply:
I've tried a number of permutations including moving this data outside the function, changing the `[]struct...` to a `[...]struct{...` (this is probably worthwhile anyway) and doing the same with a range over the pointer to the array. None make a major significant difference (maxing out at 2% improvement on bytes allocated but not changing number of allocations and making virtually no difference to time performance - max of ~2% here too, though this is statistically significant).

Reply:
This is the combination of all of them: move the data segments outside the function, make pade a `[...]struct` and range over the pointer to it.

```
name       old time/op    new time/op    delta
Exp10-8      18.6µs ± 5%    18.7µs ± 6%    ~     (p=0.171 n=10+10)
Exp100-8     3.85ms ± 3%    3.80ms ± 1%  -1.34%  (p=0.011 n=10+10)
Exp1000-8     1.99s ± 1%     2.00s ± 1%  +0.64%  (p=0.001 n=10+9)

name       old alloc/op   new alloc/op   delta
Exp10-8      1.41kB ± 2%    1.41kB ± 2%    ~     (p=1.000 n=10+10)
Exp100-8     95.6kB ± 2%    93.2kB ± 1%  -2.44%  (p=0.000 n=9+10)
Exp1000-8    91.6MB ± 0%    91.6MB ± 0%    ~     (p=0.724 n=8+9)

name       old allocs/op  new allocs/op  delta
Exp10-8        18.6 ± 8%      18.6 ± 8%    ~     (p=1.000 n=10+10)
Exp100-8       49.0 ± 0%      49.0 ± 0%    ~     (all equal)
Exp1000-8       140 ± 4%       141 ± 6%    ~     (p=0.992 n=10+10)
```

I think we just leave it as is.

Reply:
SGTM